cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu

```diff
@@ -39,553 +42,369 @@
-        printf("  GPU %2d(%s): %4.0f GF/s  %i Celsius\n", (int)i,clientFaulty.at(i) ? "FAULTY" : "OK", clientGflops.at(i), clientTemp.at(i));
+    // Create one thread per device - burn is still off here.
+    for (int i = 0; i < devCount; i++)
+    {
+        BurnTracker * bt = new (&trackThreads[i]) BurnTracker();
+        threads.push_back(std::thread(startBurn<T>,
+                                      i, smi_handle,
+                                      A, B,
+                                      bt
+                          )
+        );
     }
-    printf("\n");
-}
-
-template<class T> void launch(int runLength, bool useDoubles) {
-    //    std::system("nvidia-smi -L");
 
-    // Initializing A and B with random data
-    T *A = (T*) malloc(sizeof(T)*SIZE*SIZE);
-    T *B = (T*) malloc(sizeof(T)*SIZE*SIZE);
-    srand(10);
-    for (size_t i = 0; i < SIZE*SIZE; ++i) {
-        A[i] = (T)((double)(rand()%1000000)/100000.0);
-        B[i] = (T)((double)(rand()%1000000)/100000.0);
+    // Hold until all the threads are done with the init.
+    {
+        std::unique_lock<std::mutex> lk(cv_m);
+        cv.wait(lk, []{return startUpCounter == devCount;});
     }
 
-    // Forking a process..  This one checks the number of devices to use,
-    // returns the value, and continues to use the first one.
-    int mainPipe[2];
-    pipe(mainPipe);
-    int readMain = mainPipe[0];
-    std::vector<int> clientPipes;
-    std::vector<pid_t> clientPids;
-    clientPipes.push_back(readMain);
-
-    pid_t myPid = fork();
-    if (!myPid) {
-        // Child
-        close(mainPipe[0]);
-        int writeFd = mainPipe[1];
-        int devCount = initCuda();
-        write(writeFd, &devCount, sizeof(int));
-        startBurn<T>(0, writeFd, A, B, useDoubles);
-        close(writeFd);
-        return;
-    }
-    else {
-        clientPids.push_back(myPid);
-        close(mainPipe[1]);
-        int devCount;
-        read(readMain, &devCount, sizeof(int));
-
-        if (!devCount) {
-            fprintf(stderr, "No CUDA devices\n");
-        }
-        else {
-            for (int i = 1; i < devCount; ++i) {
-                int slavePipe[2];
-                pipe(slavePipe);
-                clientPipes.push_back(slavePipe[0]);
-                pid_t slavePid = fork();
-                if (!slavePid) {
-                    // Child
-                    close(slavePipe[0]);
-                    initCuda();
-                    startBurn<T>(i, slavePipe[1], A, B, useDoubles);
-                    close(slavePipe[1]);
-                    return;
-                }
-                else {
-                    clientPids.push_back(slavePid);
-                    close(slavePipe[1]);
-                }
-            }
-            listenClients(clientPipes, clientPids, runLength);
-        }
+    // Burn-time.
+    burn = true;
+    cv.notify_all();
+    std::this_thread::sleep_for( std::chrono::seconds(duration) );
```

Comment:
What's the purpose of this sleep?

Reply:
That is what sets the duration of the burn time. After that sleep, the main thread sets `burn=false` which triggers the work threads to finish up their work.